File: app/src/main/java/com/damhoe/scoresheetskat/score/adapter/in/ui/ScoreViewModelFactory.java
package com.damhoe.scoresheetskat.score.adapter.in.ui;

import androidx.annotation.NonNull;
import androidx.lifecycle.ViewModel;
import androidx.lifecycle.ViewModelProvider;
import androidx.lifecycle.viewmodel.CreationExtras;

import com.damhoe.scoresheetskat.base.Result;
import com.damhoe.scoresheetskat.score.application.ports.in.CUDScoreUseCase;
import com.damhoe.scoresheetskat.score.application.ports.in.LoadScoreUseCase;
import com.damhoe.scoresheetskat.score.domain.ScoreRequest;
import com.damhoe.scoresheetskat.score.domain.SkatScore;

import javax.inject.Inject;

public class ScoreViewModelFactory implements ViewModelProvider.Factory {

   private final CUDScoreUseCase cudScoreUseCase;
   private final LoadScoreUseCase loadScoreUseCase;
   private ScoreRequest scoreRequest;

   @Inject
   public ScoreViewModelFactory(CUDScoreUseCase cudScoreUseCase,
                                LoadScoreUseCase loadScoreUseCase) {
      this.cudScoreUseCase = cudScoreUseCase;
      this.loadScoreUseCase = loadScoreUseCase;
   }

   public void setScoreRequest(ScoreRequest scoreRequest) {
      this.scoreRequest = scoreRequest;
   }

   @NonNull
   @Override
   @SuppressWarnings("unchecked")
   public <T extends ViewModel> T create(@NonNull Class<T> modelClass) {
      if (modelClass != ScoreViewModel.class) {
         throw new IllegalArgumentException("Only ScoreViewModel allowed in factory method.");
      }

      if (scoreRequest.getScoreId() == -1L) {
         return (T) new ScoreViewModel.Builder(cudScoreUseCase)
                 .setRequest(scoreRequest)
                 .build();
      }

      Result<SkatScore> scoreResult = loadScoreUseCase.loadScore(scoreRequest.getScoreId());

      if (scoreResult.isSuccess()) {
         return (T) new ScoreViewModel.Builder(cudScoreUseCase)
                 .setRequest(scoreRequest)
                 .fromScore(scoreResult.getValue())
                 .build();
      }

      throw new RuntimeException(String.format("Failed to load the score for id: %s Message: %s",
                      scoreRequest.getScoreId(), scoreResult.getMessage()));
   }

   @NonNull
   @Override
   @SuppressWarnings("unchecked")
   public <T extends ViewModel> T create(@NonNull Class<T> modelClass, @NonNull CreationExtras extras) {
      if (modelClass != ScoreViewModel.class) {
         throw new IllegalArgumentException("Only ScoreViewModel allowed in factory method.");
      }

      if (scoreRequest.getScoreId() == -1) {
         return (T) new ScoreViewModel.Builder(cudScoreUseCase)
                 .setRequest(scoreRequest)
                 .build();
      }

      Result<SkatScore> scoreResult = loadScoreUseCase.loadScore(scoreRequest.getScoreId());

      if (scoreResult.isSuccess()) {
         return (T) new ScoreViewModel.Builder(cudScoreUseCase)
                 .setRequest(scoreRequest)
                 .fromScore(scoreResult.getValue())
                 .build();
      }

      throw new RuntimeException(String.format("Failed to load the score for id: %s Message: %s",
              scoreRequest.getScoreId(), scoreResult.getMessage()));
   }
}

File: app/src/main/java/com/damhoe/scoresheetskat/score/adapter/in/ui/ScoreViewModelFactory.kt
package com.damhoe.scoresheetskat.score.adapter.`in`.ui

import androidx.lifecycle.ViewModel
import androidx.lifecycle.ViewModelProvider
import androidx.lifecycle.viewmodel.CreationExtras
import com.damhoe.scoresheetskat.score.application.ports.`in`.CreateScoreUseCase
import com.damhoe.scoresheetskat.score.application.ports.`in`.GetScoreUseCase
import com.damhoe.scoresheetskat.score.domain.ScoreRequest
import com.damhoe.scoresheetskat.score.domain.SkatScore
import javax.inject.Inject

class ScoreViewModelFactory @Inject constructor(
    private val createScoreUseCase: CreateScoreUseCase,
    private val getScoreUseCase: GetScoreUseCase
) : ViewModelProvider.Factory {

    private lateinit var scoreRequest: ScoreRequest

    fun setScoreRequest(scoreRequest: ScoreRequest) {
        this.scoreRequest = scoreRequest
    }

    @Suppress("UNCHECKED_CAST")
    override fun <T : ViewModel> create(modelClass: Class<T>): T {
        require(modelClass == ScoreViewModel::class.java) {
            "Only ScoreViewModel allowed in factory method."
        }

        return ScoreViewModel.Builder(createScoreUseCase)
            .setRequest(scoreRequest)
            .apply {
                if (scoreRequest.scoreId != -1L) {
                    getScoreUseCase.getScore(scoreRequest.scoreId).onSuccess {
                        fromScore(it)
                    }
                }
            }
            .build() as T
    }


    @Suppress("UNCHECKED_CAST")
    override fun <T : ViewModel> create(modelClass: Class<T>, extras: CreationExtras): T {
        require(modelClass == ScoreViewModel::class.java) {
            "Only ScoreViewModel allowed in factory method."
        }

        return ScoreViewModel.Builder(createScoreUseCase)
            .setRequest(scoreRequest)
            .apply {
                if (scoreRequest.scoreId != -1L) {
                    getScoreUseCase.getScore(scoreRequest.scoreId).onSuccess {
                        fromScore(it)
                    }
                }
            }
            .build() as T
    }
}